### DISCUSSION TITLE
Can you add a source in IModelContentChangedEvent?

### DISCUSSION THREAD

**Discussion by InfinityDJKing** [microsoft/monaco-editor#3640] [Q&A]
like ICursorSelectionChangedEvent
I want to filter event and do code sync via websocket
Now the change event will trigger a edit event dead loop

**Comment by hediet** [comment]
There is no end-to-end source, however it is tracked if the change was caused by an undo or redo operation.
For syncing, I recommend setting and clearing a flag while you apply your change. Then, in the change event listener, you can check if that flag is set.

**Comment by InfinityDJKing** [comment]
> There is no end-to-end source, however it is tracked if the change was caused by an undo or redo operation.
> For syncing, I recommend setting and clearing a flag while you apply your change. Then, in the change event listener, you can check if that flag is set.

Could you give me some examples to show me how to add a flag to an event?
I use 'IStandaloneCodeEditor.setValue' and 'ITextModel.pushEditOperations' to set code change.
How can I add some flags into this event?
Thanks!

**Comment by hediet** [comment]
Don't attach the flag to event, but rather store it in a global/instance variable:
```ts
public doUpdate() {
    this.updating = true;
    try {
        model.setValue(...);
    } finally {
        this.updating = false;
    }
}

public handleModelContentChange(e) {
    if (this.updating) {
        return;
    }
    // ...
}
```

Events are handled synchronously.
